src/wazuh_db/wdb_global.c

```diff
@@ -1415,6 +1415,251 @@ cJSON* wdb_global_get_agents_by_connection_status (wdb_t *wdb, int last_agent_id
+            return OS_INVALID;
+        }
+    }
+
+    if (wdb_stmt_cache(wdb, WDB_STMT_GLOBAL_VACUUM_INTO) < 0) {
+        mdebug1("Cannot cache statement");
+        snprintf(output, OS_MAXSTR + 1, "Cannot cache statement");
+        return OS_INVALID;
+    }
+
+    stmt = wdb->stmt[WDB_STMT_GLOBAL_VACUUM_INTO];
+
+    if (sqlite3_bind_text(stmt, 1, path , -1, NULL) != SQLITE_OK) {
+        merror("DB(%s) sqlite3_bind_text(): %s", wdb->id, sqlite3_errmsg(wdb->db));
+        snprintf(output, OS_MAXSTR + 1, "DB(%s) sqlite3_bind_text(): %s", wdb->id, sqlite3_errmsg(wdb->db));
+        return OS_INVALID;
+    }
+
+    switch (wdb_step(stmt)) {
+    case SQLITE_ROW:
+    case SQLITE_DONE:
+        result = OS_SUCCESS;
+        break;
+    default:
+        mdebug1("SQLite: %s", sqlite3_errmsg(wdb->db));
+        snprintf(output, OS_MAXSTR + 1, "SQLite: %s", sqlite3_errmsg(wdb->db));
+        result = OS_INVALID;
+    }
+
+    if(OS_SUCCESS == result) {
+        snprintf(path_compressed, PATH_MAX, "%s.gz", path);
+        result = w_compress_gzfile(path, path_compressed);
+        if(OS_SUCCESS == result) {
+            unlink(path);
+            wdb_remove_old_backup();
+        } else {
+            snprintf(output, OS_MAXSTR + 1, "Failed during database backup compression");
+        }
+    }
+
+    return result;
+}
+
+int wdb_remove_old_backup() {
+    DIR* dp = NULL;
+    struct dirent *entry = NULL;
+    time_t current_time = 0;
+    struct stat backup_info = {0};
+    int number_of_files = 0;
+    char tmp_path[OS_SIZE_512] = {0};
+    char** backups_array = NULL;
+    time_t* backups_time_array = NULL;
+    int backups_to_delete = 0;
+
+    dp = opendir(WDB_BACKUP_FOLDER);
+
+    if(!dp) {
+        mdebug1("Unable to open backup directory '%s'", WDB_BACKUP_FOLDER);
+        return OS_INVALID;
+    }
+
+    current_time = time(NULL);
+
+    // The pre-restore snapshot is not considered for the max_files limit
+    while (entry = readdir(dp), entry) {
+        if ((strcmp(entry->d_name, ".") == 0) || (strcmp(entry->d_name, "..") == 0) ||  (strcmp(entry->d_name, "global.db-backup-pre_restore.gz") == 0)) {
+            continue;
+        }
+
+        number_of_files++;
+        os_realloc(backups_array, number_of_files * sizeof(char*), backups_array);
+        os_realloc(backups_time_array, number_of_files * sizeof(time_t), backups_time_array);
+
+        snprintf(tmp_path, OS_SIZE_512, "%s/%s", WDB_BACKUP_FOLDER, entry->d_name);
+        if(!stat(tmp_path, &backup_info)) {
+            os_strdup(tmp_path, backups_array[number_of_files-1]);
+            backups_time_array[number_of_files-1] = current_time - backup_info.st_mtime;
+        } else {
+            backups_array[number_of_files-1] = NULL;
+            backups_time_array[number_of_files-1] = -1;
+        }
+    }
+
+    backups_to_delete = number_of_files - wconfig.wdb_backup_settings[0].max_files;
```

Comment:
I think we can simplify the logic creating an ordered list of names and stat_time during the read of all the files.
Doing so, we just need to step in the <wconfig.wdb_backup_settings[0].max_files> element and start removing after that.

Reply:
The second approach was implemented !